File: sdk/kotlin/src/main/kotlin/com/loomup/client/OfflineStore.kt
package com.loomup.client

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.sql.DriverManager
import java.util.UUID

interface SyncStorage {
    suspend fun getItem(key: String): String?
    suspend fun setItem(key: String, value: String)
    suspend fun removeItem(key: String)
}

class MemorySyncStorage : SyncStorage {
    private val values = mutableMapOf<String, String>()
    override suspend fun getItem(key: String) = values[key]
    override suspend fun setItem(key: String, value: String) { values[key] = value }
    override suspend fun removeItem(key: String) { values.remove(key) }
}

/** JVM/desktop SQLite persistence; Android can provide the same SyncStorage contract via Room. */
class SQLiteSyncStorage(database: File) : SyncStorage, AutoCloseable {
    private val connection = DriverManager.getConnection("jdbc:sqlite:${database.absolutePath}")
    init { connection.createStatement().use { it.execute("CREATE TABLE IF NOT EXISTS loomup_sync_store (key TEXT PRIMARY KEY, value TEXT NOT NULL, updated_at INTEGER NOT NULL)") } }
    override suspend fun getItem(key: String): String? = connection.prepareStatement("SELECT value FROM loomup_sync_store WHERE key=?").use { statement ->
        statement.setString(1, key); statement.executeQuery().use { if (it.next()) it.getString(1) else null }
    }
    override suspend fun setItem(key: String, value: String) {
        connection.prepareStatement("INSERT INTO loomup_sync_store(key,value,updated_at) VALUES(?,?,?) ON CONFLICT(key) DO UPDATE SET value=excluded.value,updated_at=excluded.updated_at").use {
            it.setString(1, key); it.setString(2, value); it.setLong(3, System.currentTimeMillis() / 1000); it.executeUpdate()
        }
    }
    override suspend fun removeItem(key: String) { connection.prepareStatement("DELETE FROM loomup_sync_store WHERE key=?").use { it.setString(1, key); it.executeUpdate() } }
    override fun close() = connection.close()
}

enum class OfflinePhase { IDLE, SYNCING, OFFLINE, CONFLICT, ERROR }
data class OfflineStatus(val phase: OfflinePhase, val online: Boolean, val cursor: Long, val pending: Int, val conflicts: Int, val lastError: String? = null)
@Serializable data class OfflineConflict(val mutation: SyncMutation, val error: SyncMutationError)
@Serializable private data class LocalRecord(val data: Map<String, JsonValue>, val version: Long)
@Serializable private data class PersistedOfflineState(
    val format: Int = 1,
    val clientId: String = UUID.randomUUID().toString(),
    var schemaVersion: String = "",
    var cursor: Long = 0,
    val rows: MutableMap<String, MutableMap<String, LocalRecord>> = mutableMapOf(),
    val pending: MutableList<SyncMutation> = mutableListOf(),
    val conflicts: MutableList<OfflineConflict> = mutableListOf(),
)

class OfflineStore private constructor(
    private val transport: SyncTransport,
    private val resources: List<String>,
    private val storage: SyncStorage,
    private val storageKey: String,
    private val primaryKeys: Map<String, String>,
    private var state: PersistedOfflineState,
    online: Boolean,
) {
    private val mutex = Mutex()
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private var online = online
    private var phase = if (online) OfflinePhase.IDLE else OfflinePhase.OFFLINE
    private var lastError: String? = null
    private val _status = MutableStateFlow(snapshot())
    val status: StateFlow<OfflineStatus> = _status
    private val realtimeScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private var realtimeUnsubscribes: List<Unsubscribe> = emptyList()

    private fun snapshot() = OfflineStatus(phase, online, state.cursor, state.pending.size, state.conflicts.size, lastError)
    private fun notifyStatus() { _status.value = snapshot() }
    private suspend fun persist() = storage.setItem(storageKey, json.encodeToString(state))
    private fun pk(resource: String) = primaryKeys[resource] ?: "id"
    private fun requireResource(resource: String) = require(resources.contains(resource)) { "resource $resource is not synchronized" }

    suspend fun find(resource: String): List<Map<String, JsonValue>> = mutex.withLock { requireResource(resource); state.rows[resource]?.values?.map { it.data }.orEmpty() }
    suspend fun get(resource: String, id: String): Map<String, JsonValue>? = mutex.withLock { requireResource(resource); state.rows[resource]?.get(id)?.data }
    suspend fun conflicts(): List<OfflineConflict> = mutex.withLock { state.conflicts.toList() }

    suspend fun create(resource: String, data: Map<String, JsonValue>, recordId: String? = null, mutationId: String = UUID.randomUUID().toString()): Map<String, JsonValue> = mutex.withLock {
        requireResource(resource); val id = recordId ?: data[pk(resource)]?.stringValue ?: UUID.randomUUID().toString()
        val row = data + (pk(resource) to JsonValue.String(id)); state.rows.getOrPut(resource) { mutableMapOf() }[id] = LocalRecord(row, 0)
        state.pending += SyncMutation(mutationId, resource, "create", id, row); persist(); notifyStatus(); if (online) syncLocked(); row
    }
    suspend fun update(resource: String, id: String, patch: Map<String, JsonValue>, mutationId: String = UUID.randomUUID().toString()): Map<String, JsonValue> = mutex.withLock {
        requireResource(resource); val old = state.rows[resource]?.get(id) ?: error("local record not found"); val row = old.data + patch
        state.rows[resource]!![id] = LocalRecord(row, old.version); state.pending += SyncMutation(mutationId, resource, "update", id, patch, old.version)
        persist(); notifyStatus(); if (online) syncLocked(); row
    }
    suspend fun remove(resource: String, id: String, mutationId: String = UUID.randomUUID().toString()) = mutex.withLock {
        requireResource(resource); val old = state.rows[resource]?.remove(id) ?: error("local record not found")
        state.pending += SyncMutation(mutationId, resource, "delete", id, baseSequence = old.version); persist(); notifyStatus(); if (online) syncLocked()
    }
    suspend fun setOnline(value: Boolean) = mutex.withLock { online = value; phase = if (value) OfflinePhase.IDLE else OfflinePhase.OFFLINE; notifyStatus(); if (value) syncLocked() }
    suspend fun sync() = mutex.withLock { syncLocked() }

    /** Pull whenever Loomup realtime reports that a synchronized resource changed. */
    fun startRealtime(client: LoomupClient) {
        stopRealtime()
        realtimeUnsubscribes = resources.map { resource ->
            client.from(resource).subscribe { realtimeScope.launch { sync() } }
        }
    }
    fun stopRealtime() { realtimeUnsubscribes.forEach { it() }; realtimeUnsubscribes = emptyList() }
    fun close() { stopRealtime(); realtimeScope.cancel() }

    private suspend fun syncLocked() {
        if (!online) { phase = OfflinePhase.OFFLINE; notifyStatus(); return }
        phase = OfflinePhase.SYNCING; lastError = null; notifyStatus()
        try {
            if (state.schemaVersion.isEmpty()) bootstrap()
            while (state.pending.isNotEmpty()) {
                val mutation = state.pending.first(); val result = transport.syncMutations(listOf(mutation)).results.first()
                if (result.status == "acknowledged" && result.sequence != null) {
                    state.pending.removeAt(0); mutation.recordId?.let { id -> if (mutation.operation == "delete") state.rows[mutation.resource]?.remove(id) else result.record?.let { state.rows.getOrPut(mutation.resource) { mutableMapOf() }[id] = LocalRecord(it, result.sequence) } }
                } else if (result.status == "conflict" || result.status == "rejected") {
                    state.pending.removeAt(0); state.conflicts += OfflineConflict(mutation, result.error ?: SyncMutationError("conflict", "mutation rejected")); break
                } else break
            }
            if (state.conflicts.isEmpty()) pullAll(); persist(); phase = if (state.conflicts.isEmpty()) OfflinePhase.IDLE else OfflinePhase.CONFLICT
        } catch (error: LoomupError) {
            if (error.code == "reset_required") try { bootstrap(); persist(); phase = if (state.conflicts.isEmpty()) OfflinePhase.IDLE else OfflinePhase.CONFLICT }
            catch (resetError: Exception) { phase = OfflinePhase.ERROR; lastError = resetError.message ?: resetError.toString() }
            else { phase = OfflinePhase.ERROR; lastError = error.message ?: error.toString() }
        } catch (error: Exception) { phase = OfflinePhase.ERROR; lastError = error.message ?: error.toString() }
        notifyStatus()
    }
    private suspend fun bootstrap() {
        val response = transport.syncBootstrap(resources, state.clientId); state.rows.clear()
        resources.forEach { resource -> response.resources[resource]?.records.orEmpty().forEach { record -> record.data[pk(resource)]?.stringValue?.let { state.rows.getOrPut(resource) { mutableMapOf() }[it] = LocalRecord(record.data, record.version) } } }
        state.cursor = response.cursor; state.schemaVersion = response.schemaVersion; applyOptimisticPending()
    }
    private fun applyOptimisticPending() {
        state.pending.forEach { mutation -> mutation.recordId?.let { id ->
            when (mutation.operation) {
                "delete" -> state.rows[mutation.resource]?.remove(id)
                "create" -> mutation.data?.let { state.rows.getOrPut(mutation.resource) { mutableMapOf() }[id] = LocalRecord(it, 0) }
                "update" -> mutation.data?.let { patch -> state.rows[mutation.resource]?.get(id)?.let { state.rows[mutation.resource]!![id] = LocalRecord(it.data + patch, it.version) } }
            }
        } }
    }
    private suspend fun pullAll() { do { val response = transport.syncPull(state.cursor, resources, state.clientId); response.events.forEach { event -> val current = state.rows[event.resource]?.get(event.recordId); if (current == null || current.version < event.sequence) { if (event.operation == "DELETE" || event.after == null) state.rows[event.resource]?.remove(event.recordId) else state.rows.getOrPut(event.resource) { mutableMapOf() }[event.recordId] = LocalRecord(event.after, event.sequence) } }; state.cursor = response.cursor; state.schemaVersion = response.schemaVersion } while (response.hasMore) }

    companion object {
        suspend fun open(transport: SyncTransport, resources: List<String>, storage: SyncStorage = MemorySyncStorage(), storageKey: String = "loomup.sync.v1", primaryKeys: Map<String, String> = emptyMap(), online: Boolean = true): OfflineStore {
            require(resources.isNotEmpty()); val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }; val saved = storage.getItem(storageKey)
            val state = saved?.let { runCatching { json.decodeFromString<PersistedOfflineState>(it) }.getOrNull() } ?: PersistedOfflineState()
            return OfflineStore(transport, resources.distinct().sorted(), storage, storageKey, primaryKeys, state, online).also { if (online) it.sync() }
        }
    }
}

suspend fun LoomupClient.offline(resources: List<String>, storage: SyncStorage = MemorySyncStorage(), storageKey: String = "loomup.sync.v1", primaryKeys: Map<String, String> = emptyMap(), online: Boolean = true): OfflineStore =
    OfflineStore.open(LoomupSyncTransport(this), resources, storage, storageKey, primaryKeys, online).also { it.startRealtime(this) }

File: swift/Sources/Loomup/OfflineStore.swift
import Foundation

public protocol SyncStorage: Sendable {
    func getItem(_ key: String) async throws -> Data?
    func setItem(_ key: String, value: Data) async throws
    func removeItem(_ key: String) async throws
}

public actor MemorySyncStorage: SyncStorage {
    private var values: [String: Data] = [:]
    public init() {}
    public func getItem(_ key: String) -> Data? { values[key] }
    public func setItem(_ key: String, value: Data) { values[key] = value }
    public func removeItem(_ key: String) { values.removeValue(forKey: key) }
}

public struct OfflineStatus: Sendable, Equatable {
    public enum Phase: String, Sendable { case idle, syncing, offline, conflict, error }
    public var phase: Phase
    public var online: Bool
    public var cursor: Int64
    public var pending: Int
    public var conflicts: Int
    public var lastError: String?
}

public struct OfflineConflict: Codable, Sendable, Equatable {
    public var mutation: SyncMutation
    public var error: SyncMutationError
}

private struct LocalRecord: Codable, Sendable {
    var data: [String: JSONValue]
    var version: Int64
}

private struct OfflineState: Codable, Sendable {
    var format = 1
    var clientId = UUID().uuidString
    var schemaVersion = ""
    var cursor: Int64 = 0
    var rows: [String: [String: LocalRecord]] = [:]
    var pending: [SyncMutation] = []
    var conflicts: [OfflineConflict] = []
}

public actor OfflineStore {
    private let transport: SyncTransport
    private let resources: [String]
    private let storage: SyncStorage
    private let storageKey: String
    private let primaryKeys: [String: String]
    private var state: OfflineState
    private var phase: OfflineStatus.Phase
    private var online: Bool
    private var lastError: String?
    private var continuations: [UUID: AsyncStream<OfflineStatus>.Continuation] = [:]
    private var realtimeUnsubscribes: [Unsubscribe] = []

    private init(
        transport: SyncTransport,
        resources: [String],
        storage: SyncStorage,
        storageKey: String,
        primaryKeys: [String: String],
        state: OfflineState,
        online: Bool
    ) {
        self.transport = transport
        self.resources = resources
        self.storage = storage
        self.storageKey = storageKey
        self.primaryKeys = primaryKeys
        self.state = state
        self.online = online
        phase = online ? .idle : .offline
    }

    public static func open(
        transport: SyncTransport,
        resources: [String],
        storage: SyncStorage = MemorySyncStorage(),
        storageKey: String = "loomup.sync.v1",
        primaryKeys: [String: String] = [:],
        online: Bool = true
    ) async throws -> OfflineStore {
        guard !resources.isEmpty else { throw LoomupError("OfflineStore requires resources") }
        let saved = try await storage.getItem(storageKey)
        let state = saved.flatMap { try? JSONDecoder().decode(OfflineState.self, from: $0) } ?? OfflineState()
        let store = OfflineStore(
            transport: transport, resources: Array(Set(resources)).sorted(), storage: storage,
            storageKey: storageKey, primaryKeys: primaryKeys, state: state, online: online
        )
        if online { await store.sync() }
        return store
    }

    public var status: OfflineStatus {
        OfflineStatus(
            phase: phase, online: online, cursor: state.cursor, pending: state.pending.count,
            conflicts: state.conflicts.count, lastError: lastError
        )
    }

    public var conflicts: [OfflineConflict] { state.conflicts }

    public func statusStream() -> AsyncStream<OfflineStatus> {
        let id = UUID()
        return AsyncStream { continuation in
            continuations[id] = continuation
            continuation.yield(status)
            continuation.onTermination = { [weak self] _ in Task { await self?.removeContinuation(id) } }
        }
    }

    private func removeContinuation(_ id: UUID) { continuations.removeValue(forKey: id) }
    private func notify() { for continuation in continuations.values { continuation.yield(status) } }
    private func persist() async throws { try await storage.setItem(storageKey, value: JSONEncoder().encode(state)) }
    private func pk(_ resource: String) -> String { primaryKeys[resource] ?? "id" }
    private func require(_ resource: String) throws {
        if !resources.contains(resource) { throw LoomupError("resource \(resource) is not synchronized") }
    }

    public func find(_ resource: String) throws -> [[String: JSONValue]] {
        try require(resource)
        return (state.rows[resource] ?? [:]).values.map(\.data)
    }

    public func get(_ resource: String, id: String) throws -> [String: JSONValue]? {
        try require(resource)
        return state.rows[resource]?[id]?.data
    }

    @discardableResult
    public func create(
        _ resource: String,
        data: [String: JSONValue],
        recordId: String? = nil,
        mutationId: String = UUID().uuidString
    ) async throws -> [String: JSONValue] {
        try require(resource)
        let id = recordId ?? data[pk(resource)]?.stringValue ?? UUID().uuidString
        var row = data; row[pk(resource)] = .string(id)
        state.rows[resource, default: [:]][id] = LocalRecord(data: row, version: 0)
        state.pending.append(SyncMutation(id: mutationId, resource: resource, operation: "create", recordId: id, data: row))
        try await persist(); notify()
        if online { await sync() }
        return row
    }

    @discardableResult
    public func update(
        _ resource: String, id: String, patch: [String: JSONValue],
        mutationId: String = UUID().uuidString
    ) async throws -> [String: JSONValue] {
        try require(resource)
        guard var record = state.rows[resource]?[id] else { throw LoomupError("local record not found") }
        record.data.merge(patch) { _, new in new }
        state.rows[resource]?[id] = record
        state.pending.append(SyncMutation(
            id: mutationId, resource: resource, operation: "update", recordId: id,
            data: patch, baseSequence: record.version
        ))
        try await persist(); notify()
        if online { await sync() }
        return record.data
    }

    public func remove(
        _ resource: String, id: String, mutationId: String = UUID().uuidString
    ) async throws {
        try require(resource)
        guard let record = state.rows[resource]?[id] else { throw LoomupError("local record not found") }
        state.rows[resource]?.removeValue(forKey: id)
        state.pending.append(SyncMutation(
            id: mutationId, resource: resource, operation: "delete", recordId: id,
            baseSequence: record.version
        ))
        try await persist(); notify()
        if online { await sync() }
    }

    public func setOnline(_ value: Bool) async {
        online = value; phase = value ? .idle : .offline; notify()
        if value { await sync() }
    }

    public func sync() async {
        guard online else { phase = .offline; notify(); return }
        phase = .syncing; lastError = nil; notify()
        do {
            if state.schemaVersion.isEmpty { try await bootstrap() }
            while let mutation = state.pending.first {
                let response = try await transport.syncMutations([mutation])
                guard let result = response.results.first else { throw LoomupError("empty mutation response") }
                if result.status == "acknowledged", let sequence = result.sequence {
                    state.pending.removeFirst()
                    if let id = mutation.recordId {
                        if mutation.operation == "delete" { state.rows[mutation.resource]?.removeValue(forKey: id) }
                        else if let row = result.record { state.rows[mutation.resource, default: [:]][id] = LocalRecord(data: row, version: sequence) }
                    }
                } else if result.status == "conflict" || result.status == "rejected" {
                    state.pending.removeFirst()
                    state.conflicts.append(OfflineConflict(
                        mutation: mutation,
                        error: result.error ?? SyncMutationError(code: "conflict", message: "mutation rejected")
                    ))
                    break
                } else { break }
            }
            if state.conflicts.isEmpty { try await pullAll() }
            try await persist()
            phase = state.conflicts.isEmpty ? .idle : .conflict
        } catch let error as LoomupError where error.code == "reset_required" {
            do {
                try await bootstrap()
                try await persist()
                phase = state.conflicts.isEmpty ? .idle : .conflict
            } catch {
                phase = .error; lastError = String(describing: error)
            }
        } catch {
            phase = .error; lastError = String(describing: error)
        }
        notify()
    }

    /// Pull whenever Loomup realtime reports that a synchronized resource changed.
    public func startRealtime(client: LoomupClient) {
        stopRealtime()
        realtimeUnsubscribes = resources.map { resource in
            client.from(resource).subscribe { [weak self] _ in
                Task { await self?.sync() }
            }
        }
    }

    public func stopRealtime() {
        realtimeUnsubscribes.forEach { $0() }
        realtimeUnsubscribes.removeAll()
    }

    public func close() {
        stopRealtime()
        continuations.values.forEach { $0.finish() }
        continuations.removeAll()
    }

    private func bootstrap() async throws {
        let response = try await transport.syncBootstrap(resources: resources, clientId: state.clientId)
        var rows: [String: [String: LocalRecord]] = [:]
        for resource in resources {
            for record in response.resources[resource]?.records ?? [] {
                guard let id = record.data[pk(resource)]?.stringValue else { continue }
                rows[resource, default: [:]][id] = LocalRecord(data: record.data, version: record.version)
            }
        }
        state.rows = rows; state.cursor = response.cursor; state.schemaVersion = response.schemaVersion
        applyOptimisticPending()
    }

    private func applyOptimisticPending() {
        for mutation in state.pending {
            guard let id = mutation.recordId else { continue }
            if mutation.operation == "delete" { state.rows[mutation.resource]?.removeValue(forKey: id); continue }
            if mutation.operation == "create", let data = mutation.data {
                state.rows[mutation.resource, default: [:]][id] = LocalRecord(data: data, version: 0)
            } else if mutation.operation == "update", let patch = mutation.data, var record = state.rows[mutation.resource]?[id] {
                record.data.merge(patch) { _, new in new }; state.rows[mutation.resource]?[id] = record
            }
        }
    }

    private func pullAll() async throws {
        while true {
            let response = try await transport.syncPull(cursor: state.cursor, resources: resources, clientId: state.clientId)
            if !state.schemaVersion.isEmpty, response.schemaVersion != state.schemaVersion { try await bootstrap(); return }
            for event in response.events {
                let current = state.rows[event.resource]?[event.recordId]
                if let current, current.version >= event.sequence { continue }
                if event.operation == "DELETE" || event.after == nil { state.rows[event.resource]?.removeValue(forKey: event.recordId) }
                else { state.rows[event.resource, default: [:]][event.recordId] = LocalRecord(data: event.after!, version: event.sequence) }
            }
            state.cursor = response.cursor; state.schemaVersion = response.schemaVersion
            if !response.hasMore { break }
        }
    }
}

public extension LoomupClient {
    func offline(
        resources: [String], storage: SyncStorage = MemorySyncStorage(), storageKey: String = "loomup.sync.v1",
        primaryKeys: [String: String] = [:], online: Bool = true
    ) async throws -> OfflineStore {
        let store = try await OfflineStore.open(
            transport: self, resources: resources, storage: storage,
            storageKey: storageKey, primaryKeys: primaryKeys, online: online
        )
        await store.startRealtime(client: self)
        return store
    }
}
